ADBDEV-5036: Fix gpstop -u not updating GUCs on segments

bandetto opened the pull request:
Fix gpstop -u not updating GUCs on segments

Backend process GUCs marked as GUC_GPDB_NEED_SYNC are overwritten by coordinator
upon receiving a new connection. The coordinator sends it's configuration to the
segments the same way as if it were a frontend communicating with the
coordinator. This means the default priority for these GUCs becomes client
priority.

Previously, when gpstop -u was executed, it sent a SIGHUP signal, prompting the
coordinator and segments to re-read their configuration files. However, the
priority of options derived from the configuration file is lower than that of
client options. This difference caused the segments to ignore overwritten
GUCs when they re-read their configuration files, and these GUCs were not
re-dispatched by the QD either. This situation resulted in a desynchronization
of GUCs that are required to be synchronized between the coordinator and the
segments.

The solution is to explicitly dispatch the changed configuration from the
coordinator to the segments, similar to how it is done when initializing the
backend process.

When the QD receives a SIGHUP signal, it reads its configuration file to
identify any GUCs that have been modified. If these GUCs are marked with the
GUC_GPDB_NEED_SYNC flag, the QD dispatches SET commands containing these
modified GUCs to the QEs. To differentiate between this type of synchronization
and user-initiated SET commands, a flag GP_OPT_SYNCHRONIZATION_SET has been
introduced to the distributed transaction options.

bandetto committed:
Dispatch SET from to QEs to sync GUCs on segments after SIGHUP

bandetto committed:
Add tests

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/66010

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1129929

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1129920

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1129922

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1129921

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1129923

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -5303,7 +5307,37 @@ PostgresMain(int argc, char *argv[],
+				foreach (lc, changed_guc_names)
+				{
+					const char *name = lfirst(lc);
+					changed_gucs = lappend(changed_gucs, find_option(name, false, DEBUG2));
```
The find_option function may return NULL. Why are you putting NULL in the list?

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -5303,7 +5307,37 @@ PostgresMain(int argc, char *argv[],
+					changed_gucs = lappend(changed_gucs, find_option(name, false, DEBUG2));
+				}
+
+				if (changed_gucs)
```
```suggestion
				if (changed_gucs != NIL)
```

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -5303,7 +5307,37 @@ PostgresMain(int argc, char *argv[],
+				list_free(changed_gucs);
+			}
+			else
+				ProcessConfigFile(PGC_SIGHUP);
```
Place the short branch first.

RekGRpth commented on src/backend/utils/misc/guc-file.l:
```diff
@@ -221,7 +237,7 @@ ProcessConfigFile(GucContext context)
 		 */
 		if (head == NULL)
-			return;
+			return NULL;
```
```suggestion
			return NIL;
```

RekGRpth commented on src/backend/utils/misc/guc-file.l:
```diff
@@ -422,7 +438,11 @@ ProcessConfigFile(GucContext context)
 					ereport(elevel,
 							(errmsg("parameter \"%s\" changed to \"%s\"",
 									item->name, item->value)));
+
```
?

RekGRpth commented on src/test/regress/output/dispatch.source:
```diff
@@ -1073,3 +1073,65 @@ reset gp_print_create_gang_time;
+-- to segments, so segments should have the same options as QD after gpstop -u.
+--
+--start_ignore
+CREATE LANGUAGE plpgsql;
```
The contents of the ignore block are redundant in the output.

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -5303,7 +5308,38 @@ PostgresMain(int argc, char *argv[],
+					send_guc_to_QE(changed_gucs);
+
+				list_free(changed_guc_names);
+				list_free(changed_gucs);
```
You can move this list_free and the other one below into the send_guc_to_QE function.

bandetto replied:
Isn't it more clear when the object is allocated and freed in the same context? I think it will be confusing to delegate the cleanup to a function.

RekGRpth commented on src/test/regress/output/dispatch.source:
```diff
@@ -1073,3 +1073,61 @@ reset gp_print_create_gang_time;
+-- Explicitly set statement_mem through gpconfig, then run gpstop -u. SHOW and
+-- show_on_segments() should return same results.
+--start_ignore
+\!gpconfig -c statement_mem -v '125MB'
```
The contents of the ignore block (here and below) are redundant in the output.

RekGRpth commented on src/test/regress/input/dispatch.source:
```diff
@@ -625,3 +625,46 @@ reset gp_print_create_gang_time;
+-- show_on_segments() should return same results.
+
+--start_ignore
+\!gpconfig -c statement_mem -v '125MB'
```
What happens and what should happen when we set different values for the coordinator and segments?
```sh
gpconfig -c max_connections -v 100 -m 10
```

bandetto replied:
Here's a comment on how it works:
https://github.com/arenadata/gpdb/blob/45918e2214007e6e9bf6a6b29f867444626486c9/src/backend/cdb/dispatcher/cdbgang.c#L385-L405

In short, on backend instantiation, the coordinator forcefully overwrites GUCs marked with `GUC_GPDB_NEED_SYNC` flag on QEs.

In case of `gpconfig -c max_connections -v 100 -m 10`, if `max_connections` is marked as `GUC_GPDB_NEED_SYNC`, on QEs it will be the same as on QD. Otherwise, it will not be sent, and it will be out of sync on QD and QEs.

bandetto committed:
Prohibit PGC_SIGHUP context for GUCs which need to be synchronized

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66592

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1157844

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1157845

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1157843

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66678

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1163674

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1163675

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1163667

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1163669

RekGRpth commented on src/test/regress/output/dispatch.source:
```diff
@@ -1073,3 +1073,122 @@ reset gp_print_create_gang_time;
+(3 rows)
+
+--start_ignore
+\!gpconfig -c statement_mem -v '50MB' -m '100MB'
```
The contents of the ignore block are redundant in the output.

bandetto replied:
> There is no output from the function in the input file, so it will be very inconvenient to look at both files to understand which output is correct for which settings. You can look at these ignore blocks as comments which say input values before the function call.
>
> Without these ignore blocks, the output file will look like a comment, a function call, it's results, then another comment, a function call and there's suddenly different results.

These blocks help to preserve the context of what's GUCs are being changed.

RekGRpth commented on src/test/regress/output/dispatch.source:
```diff
@@ -1073,3 +1073,122 @@ reset gp_print_create_gang_time;
+(3 rows)
+
+--start_ignore
+\!gpconfig -r statement_mem
```
The contents of the ignore block are redundant in the output.

RekGRpth commented on src/test/regress/input/dispatch.source:
```diff
@@ -625,3 +625,70 @@ reset gp_print_create_gang_time;
+\!gpconfig -c statement_mem -v '125MB'
+\!gpconfig -c backslash_quote -v 'safe_encoding'
+\!gpstop -u
+\c
```
I think, this line is redundant

bandetto committed:
Remove \c from tests

bandetto committed:
Ignore sync flags when restoring GUCs

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66965

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1179655

bandetto commented:
@RekGRpth If for some reason there's an error in the try block in `send_guc_to_QE()`, and we call `DisconnectAndDestroyAllGangs(false)` from the catch block, temp tables will be left in a half deleted state. Is this ok?

Repro:
```diff
diff --git a/src/backend/tcop/postgres.c b/src/backend/tcop/postgres.c
index fe0305111d4..85d47ad9b84 100644
--- a/src/backend/tcop/postgres.c
+++ b/src/backend/tcop/postgres.c
@@ -1581,6 +1581,7 @@ send_guc_to_QE(List *guc_list, bool is_restore)
                PG_TRY();
                {
                        DispatchSyncPGVariable(gconfig);
+                       elog(ERROR, "test");
                }
                PG_CATCH();
                {
```
In psql:
```
postgres=# create temporary table test(a int) distributed by (a);
CREATE TABLE
postgres=# \!gpstop -u
...
postgres=# select * from test;
ERROR:  relation not found (OID 366810)  (seg0 slice1 127.0.1.1:6002 pid=249929)
DETAIL:  This can be validly caused by a concurrent delete operation on this object.
```

Should we call `DisconnectAndDestroyAllGangs(true)` and `CheckForResetSession()` instead to finalize session reset?

RekGRpth commented:
> Should we call `DisconnectAndDestroyAllGangs(true)` and `CheckForResetSession()` instead to finalize session reset?

Maybe this is a separate task?

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66987

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1179867

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1179868

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1179864

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -1560,16 +1561,23 @@ CheckDebugDtmActionSqlCommandTag(const char *sqlCommandTag)
 		struct config_generic* gconfig = (struct config_generic *)lfirst(lc);
+
+		/* We can't SET GUCs whose context is too high. */
+		if (gconfig->context < PGC_SUSET ||
+			(!is_restore && !(gconfig->flags & GUC_GPDB_NEED_SYNC)))
```
Maybe this condition will be more understandable?
```suggestion
		if (!((gconfig->context >= PGC_SUSET) && (is_restore || (gconfig->flags & GUC_GPDB_NEED_SYNC)))
```

RekGRpth replied:
Or maybe this would be right?
```suggestion
		if (!(is_restore || ((gconfig->context >= PGC_SUSET) && (gconfig->flags & GUC_GPDB_NEED_SYNC)))
```

bandetto replied:
I think unchanged condition is simpler
`"If the context is too high || (this is not restore && we don't have a sync flag set)"`

RekGRpth replied:
I think `PostgresMain` should send all GUCs, not only if `gconfig->context >= PGC_SUSET`, as it was before patch.

bandetto replied:
It's impossible, since SET does not allow changing GUCs with context higher than PGC_SUSET, and we can't distinguish SET commands between sent by this function or users performing a manual SET. Before this patch, this was still the case.

RekGRpth replied:
Before patch `PostgresMain` calls `restore_guc_to_QE` and it restores all GUCs in `gp_guc_restore_list`.
After patch `PostgresMain` calls `restore_guc_to_QE` with true as `is_restore` and now it does not restore GUCS `from gp_guc_restore_list` if they `context < PGC_SUSET`.

bandetto replied:
SET command calls `set_config_option()` with `PGC_SUSET`/`PGC_USERSET`:
https://github.com/arenadata/gpdb/blob/cdc9332854a6aca15bfe98541a976342658be028/src/backend/utils/misc/guc.c#L7278-L7284
And then the context of the GUC is compared with the one passed as an argument:
https://github.com/arenadata/gpdb/blob/cdc9332854a6aca15bfe98541a976342658be028/src/backend/utils/misc/guc.c#L5928

So even if I get rid of this check, SET will not be able to restore a GUCs if it's context is too high.

RekGRpth replied:
Why not exclude continue when is_restore is true?

RekGRpth commented on src/backend/utils/misc/guc-file.l:
```diff
@@ -356,6 +374,8 @@ ProcessConfigFile(GucContext context)
 						(errmsg("parameter \"%s\" removed from configuration file, reset to default",
 								gconf->name)));
+			if (return_changed)
+				changed_gucs = lappend(changed_gucs, pstrdup(gconf->name));
```
How about returning a list of GUCs, rather than their names?
```suggestion
				changed_gucs = lappend(changed_gucs, gconf);
```

RekGRpth commented on src/backend/utils/misc/guc-file.l:
```diff
@@ -423,6 +443,9 @@ ProcessConfigFile(GucContext context)
 			}
+
+			if (return_changed)
+				changed_gucs = lappend(changed_gucs, pstrdup(item->name));
```
How about returning a list of GUCs, rather than their names?
```suggestion
				{
					struct config_generic *gconf = find_option(item->name, false, elevel);
					if (gconf != NULL)
						changed_gucs = lappend(changed_gucs, gconf);
				}
```

bandetto committed:
Construct a list of config_generic * instead of using names

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67123

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1188422

bandetto committed:
Don't continue if we are restoring GUCs

RekGRpth commented on src/backend/tcop/postgres.c:
```diff
@@ -1560,16 +1561,23 @@ CheckDebugDtmActionSqlCommandTag(const char *sqlCommandTag)
 	{
 		struct config_generic* gconfig = (struct config_generic *)lfirst(lc);
+
+		/* We can't SET GUCs whose context is too high. */
```
I think comment should be updated.

bandetto replied:
To what? It's still relevant.

RekGRpth replied:
this comment is only about condition `gconfig->context < PGC_SUSET`

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67145

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189299

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189300

bandetto committed:
Update a comment on why we don't send some GUCs

bandetto committed:
Typo

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67162

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190195

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190196

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190192

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190187

red1452 commented on src/test/regress/output/dispatch.source:
```diff
@@ -1073,3 +1073,122 @@ reset gp_print_create_gang_time;
+-----------------+-------------
+ backslash_quote | off
+ backslash_quote | off
+ backslash_quote | off
```
Value of the `backslash_quote` GUC is correctly set at all cases without fixes at the core. What should check this part of test?

bandetto replied:
Only GUCs which have `GPDB_GUC_NEED_SYNC` flag should be synchronized. `backslash_quote` is a randomly-picked setting that does not have `GPDB_GUC_NEED_SYNC` flag. This case checks that it wasn't sent to QEs, unlike `statement_mem`.

red1452 commented:
Some remarks of the description:
```
gpstop -u not updating -> gpstop -u doesn't update
The segment’s backend process GUCs marked -> backend process GUCs, which are marked
the segments as if it -> the segments the same as
some frontend that communicates -> which communicates
configuration file is lower than that client’s configuration -> than the client’s
GUCs with overwritten values are not read -> don't read
They are not re-sent either -> don't re-sent
These GUCs become desynchronized -> These GUCs are desynchronized
The solution is to explicitly re-send -> The solution is to re-send explicitly
```

bandetto committed:
Refactor tests

bandetto committed:
Fix typos in tests

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67538

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210248

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210249

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210246

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210240

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210241

RekGRpth commented:
new flaky [test](https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210240)?
```diff
DIFF FILE: ../gpdb_src/src/test/regress/regression.diffs
----------------------------------------------------------------------
--- \/home\/gpadmin\/gpdb_src\/src\/test\/regress\/expected\/dispatch\.out	2024-03-27 07:19:55.439070087 +0000
+++ \/home\/gpadmin\/gpdb_src\/src\/test\/regress\/results\/dispatch\.out	2024-03-27 07:19:55.475071263 +0000
@@ -1210,10 +1246,9 @@
 -- present. If it's not, something is terribly wrong and we were silently
 -- disconnected.
 SELECT * FROM dispatch_test_session_not_dead;
- a 
----
-(0 rows)
-
+ERROR:  relation not found (OID 67184)
+DETAIL:  This can be validly caused by a concurrent delete operation on this object.
 DROP TABLE dispatch_test_session_not_dead;
+ERROR:  table "dispatch_test_session_not_dead" does not exist
 DROP FUNCTION show_on_segments(show_name text);
 DROP ROLE dispatch_test_user2;
```

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210242

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210243

bandetto commented:
> new flaky [test](https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210240) and [test](https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1210241)?

FYI: not flaky, these are new tests that cover more cases. The method used in this patch probably would not work for synchronization, because it does not account for dispatching SET from non-superuser roles. Currently, I'm looking into the problem to see what can be done, but this patch will probably be scrapped for another approach.

bandetto committed:
Add a new transaction flag to check whether SET was for synchronization

bandetto committed:
Remove old constraints

RekGRpth commented on src/include/utils/guc.h:
```diff
@@ -804,7 +805,7 @@ extern bool gpvars_check_statement_mem(int *newval, void **extra, GucSource sour
 extern bool gpvars_check_gp_gpperfmon_send_interval(int *newval, void **extra, GucSource source);
 extern int guc_name_compare(const char *namea, const char *nameb);
-extern void DispatchSyncPGVariable(struct config_generic * gconfig);
+extern void DispatchSyncPGVariable(struct config_generic * gconfig, bool issync);
```
why `issync` here while `isSync` above?

bandetto committed:
Update comments and parameter name

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1235550

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1235551

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1235543

bandetto committed:
Update action

bandetto committed:
Add privelege check to test

RekGRpth commented:
fix abi test

bandetto committed:
Merge branch 'adb-6.x-dev' into ADBDEV-5036

bandetto committed:
Factor out changes to new functions

RekGRpth commented on src/backend/cdb/dispatcher/cdbdisp_query.c:
```diff
@@ -278,9 +282,24 @@ cdbdisp_markNamedPortalGangsDestroyed(void)
+ * isSync is true if we are QD synchronizing GUCs with QEs to bypass context
+ * constraints. See ProcessStartupPacket()
  */
 void
 CdbDispatchSetCommand(const char *strCommand, bool cancelOnError)
+{
+	return cdbdisp_dispatchSetCommandInternal(strCommand, cancelOnError, false);
+}
+
+void
+CdbDispatchSetCommandForSync(const char *strCommand)
+{
+	return cdbdisp_dispatchSetCommandInternal(strCommand, false, true);
+}
```
put these lines after cdbdisp_dispatchSetCommandInternal and you don't need redundant declaration of this function

bandetto replied:
Weren't there an internal discussion about some warnings due to static functions not having prototypes?

RekGRpth replied:
if static function is used after defined, there is no warnings/errors

RekGRpth commented on src/backend/utils/misc/guc-file.l:
```diff
@@ -130,6 +131,22 @@ STRING			\'([^'\\\n]|\\.|\'\')*\'
  */
 void
 ProcessConfigFile(GucContext context)
+{
+	(void) ProcessConfigFileImpl(context, false);
+}
+
+/*
+ * This works the same way as the ProcessConfigFile(), except that it returns a
+ * list of options (config_generic *) that were changed.
+ */
+List *
+ProcessConfigFileForSync(GucContext context)
+{
+	return ProcessConfigFileImpl(context, true);
+}
```
put these lines after ProcessConfigFileImpl and you don't need redundant declaration of this function

RekGRpth commented on src/backend/utils/misc/guc.c:
```diff
@@ -7628,7 +7648,7 @@ set_config_by_name(PG_FUNCTION_ARGS)
 		pfree(quoted_name);
 
-		CdbDispatchSetCommand(buffer.data, false /* cancelOnError */ );
+		CdbDispatchSetCommand(buffer.data, false /* cancelOnError */);
```
restore original line

RekGRpth commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -96,6 +97,9 @@ static int guc_array_compare(const void *a, const void *b);
 
 extern struct config_generic *find_option(const char *name, bool create_placeholders, int elevel);
 
+static void dispatch_sync_pg_variable(struct config_generic *gconfig,
```
this declaration is redundant

bandetto committed:
Get rid of redundant prototypes, restore a space

RekGRpth commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -5739,8 +5740,8 @@ check_gp_workfile_compression(bool *newval, void **extra, GucSource source)
-void
-DispatchSyncPGVariable(struct config_generic * gconfig)
+static void
+dispatch_sync_pg_variable(struct config_generic * gconfig, bool is_sync)
```
why not DispatchSyncPGVariableImpl or DispatchSyncPGVariableInternal?

bandetto replied:
What do you think about DispatchSyncPGVariableForSync() function name? It stands in line with other "...ForSync()" functions that were added, but sounds silly

RekGRpth replied:
Yes, it sounds strange, I also didn’t understand why cdbdisp_dispatchSetCommand**Internal** and ProcessConfigFile**Impl**.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68303

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1240843

bandetto committed:
Make names more concise, add more comments

RekGRpth commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -5739,8 +5740,8 @@ check_gp_workfile_compression(bool *newval, void **extra, GucSource source)
-void
-DispatchSyncPGVariable(struct config_generic * gconfig)
+static void
+dispatch_sync_pg_variable_internal(struct config_generic * gconfig, bool is_sync)
```
may be use word `force` instead `sync`?

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68324

bandetto committed:
Rename parameter for clarity

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68370

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1241831

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1241824

InnerLife0 commented on src/test/regress/input/dispatch.source:
```diff
@@ -625,3 +625,97 @@ reset gp_print_create_gang_time;
+-- coordinator upon receiving a new connection.
+--
+-- GUCs that were changed in the configuration file of the QD should be re-sent
+-- to segments on gpstop -u, if GUC_GPDB_NEED_SYNC is present if the flags.
```
I think it should be _present in_.

InnerLife0 commented on src/backend/cdb/cdbtm.c:
```diff
@@ -1216,6 +1222,17 @@ mppTxnOptions(bool needDtx)
 }
 
+int
+mppTxnOptionsForSync(bool needDtx, bool isSync)
```
It seems this function is used only once. Can we apply the same code without creating a new function?

bandetto replied:
GP_OPT_... macros only exist inside the cdbtm.c file. Only mppTxnOptions() and isMppTxOptions_...() functions are exported.

InnerLife0 commented on src/test/regress/input/dispatch.source:
```diff
@@ -625,3 +625,97 @@ reset gp_print_create_gang_time;
+-- Our session should not have been reset, and this table should still be
+-- present. If it's not, something is terribly wrong and we were silently
+-- disconnected.
+SELECT * FROM dispatch_test_session_not_dead;
```
I have the same successful behavior on not patched version. If it's an expected behavior, we possibly don't need this table in tests.

bandetto replied:
It's here because DisconnectAndDestroyAllGangs(false) call in the PG_CATCH() block of senc_guc_to_QE() disconnects the session. We are asserting that no error happened.

I re-read the code for the send_guc_to_QE() function, and I had a few questions for myself. Currently, potentially, some GUCs may not be sent. We just skip them and continue sending the rest of the settings. Maybe in case of an error during synchronization we should throw fatal error, or we should abort the transaction and reset the session?

bandetto replied:
> Maybe in case of an error during synchronization we should throw fatal error, or we should abort the transaction and reset the session?

Now, session resets if any of the GUCs fail to be sent. GUCs will be re-sent in the startup packet as usual.

RekGRpth replied:
> Now, session resets if any of the GUCs fail to be sent. GUCs will be re-sent in the startup packet as usual.

could you add tests for it?

bandetto replied:
> could you add tests for it?

https://github.com/arenadata/gpdb/pull/881#issuecomment-2049822909

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68643

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1259919

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68648

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1260222

bandetto committed:
Reset the session in case any of the GUCs failed to be sent

bandetto committed:
Fix typo in tests

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1260659

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1260660

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1260656

bandetto committed:
Add tests to check the failure behavior

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68671

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1262276

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1262277

bandetto committed:
Revert "Add tests to check the failure behavior"

This reverts commit a6b8ab2bf6268836024e73b8c7095e308f74e568.

bandetto committed:
Don't check a temporary table in tests

The problem with temporary table deletion was decided to be addressed in
a different PR.

bandetto commented:
After some discussion, tests for problem related to https://github.com/arenadata/gpdb/pull/881#issuecomment-2006929685 were decided to be moved to a different PR.

bandetto committed:
Revert "Reset the session in case any of the GUCs failed to be sent"

This reverts commit 1abc6d047c538a4111c48bce3e43155d767901ce.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/68895

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1274943

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1274944

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1274937

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1274938

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-5036

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/69086

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1285025

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1285026

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1285023
